File: src/dorkbox/network/dns/records/NULLRecord.java
// Copyright (c) 2004 Brian Wellington (bwelling@xbill.org)

package dorkbox.network.dns.records;

import java.io.IOException;

import dorkbox.network.dns.Compression;
import dorkbox.network.dns.DnsInput;
import dorkbox.network.dns.DnsOutput;
import dorkbox.network.dns.Name;
import dorkbox.network.dns.constants.DnsRecordType;
import dorkbox.network.dns.utils.Tokenizer;

/**
 * The NULL Record.  This has no defined purpose, but can be used to
 * hold arbitrary data.
 *
 * @author Brian Wellington
 */

public
class NULLRecord extends DnsRecord {

    private static final long serialVersionUID = -5796493183235216538L;

    private byte[] data;

    NULLRecord() {}

    @Override
    DnsRecord getObject() {
        return new NULLRecord();
    }

    @Override
    void rrFromWire(DnsInput in) throws IOException {
        data = in.readByteArray();
    }

    @Override
    void rrToWire(DnsOutput out, Compression c, boolean canonical) {
        out.writeByteArray(data);
    }

    @Override
    void rrToString(StringBuilder sb) {
        sb.append(unknownToString(data));
    }

    @Override
    void rdataFromString(Tokenizer st, Name origin) throws IOException {
        throw st.exception("no defined text format for NULL records");
    }

    /**
     * Creates a NULL record from the given data.
     *
     * @param data The contents of the record.
     */
    public
    NULLRecord(Name name, int dclass, long ttl, byte[] data) {
        super(name, DnsRecordType.NULL, dclass, ttl);

        if (data.length > 0xFFFF) {
            throw new IllegalArgumentException("data must be <65536 bytes");
        }
        this.data = data;
    }

    /**
     * Returns the contents of this record.
     */
    public
    byte[] getData() {
        return data;
    }

}

File: src/dorkbox/dns/dns/records/NULLRecord.kt
/*
 * Copyright 2021 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.dns.dns.records

import dorkbox.dns.dns.Compression
import dorkbox.dns.dns.DnsInput
import dorkbox.dns.dns.DnsOutput
import dorkbox.dns.dns.Name
import dorkbox.dns.dns.constants.DnsRecordType
import dorkbox.dns.dns.utils.Tokenizer
import java.io.IOException

/**
 * The NULL Record.  This has no defined purpose, but can be used to
 * hold arbitrary data.
 *
 * @author Brian Wellington
 */
class NULLRecord : DnsRecord {
    /**
     * Returns the contents of this record.
     */
    lateinit var data: ByteArray
        private set

    internal constructor() {}

    override val dnsRecord: DnsRecord
        get() = NULLRecord()

    @Throws(IOException::class)
    override fun rrFromWire(`in`: DnsInput) {
        data = `in`.readByteArray()
    }

    override fun rrToWire(out: DnsOutput, c: Compression?, canonical: Boolean) {
        out.writeByteArray(data)
    }

    override fun rrToString(sb: StringBuilder) {
        sb.append(unknownToString(data))
    }

    @Throws(IOException::class)
    override fun rdataFromString(st: Tokenizer, origin: Name?) {
        throw st.exception("no defined text format for NULL records")
    }

    /**
     * Creates a NULL record from the given data.
     *
     * @param data The contents of the record.
     */
    constructor(name: Name?, dclass: Int, ttl: Long, data: ByteArray) : super(name!!, DnsRecordType.NULL, dclass, ttl) {
        require(data.size <= 0xFFFF) { "data must be <65536 bytes" }
        this.data = data
    }

    companion object {
        private const val serialVersionUID = -5796493183235216538L
    }
}
